### PR TITLE
Refactor `InitializerExprTypeResolver->resolveEqualType()`

### PR DESCRIPTION
**Pull request by staabm**
move the logic into the type-classes, as requested in https://github.com/phpstan/phpstan-src/pull/2216#pullrequestreview-1286571372

### CODE REVIEW

File: src/Type/ComparisonHelper.php [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+		}
+
+		// @phpstan-ignore-next-line
+		return new ConstantBooleanType($leftType->getValue() == $rightType->getValue()); // phpcs:ignore
+	}
+
+	/**
+	 * @param callable(Type, Type): BooleanType $valueComparisonCallback
+	 */
+	public static function compareConstantArrayType(ConstantArrayType $leftType, ConstantArrayType $rightType, callable $valueComparisonCallback): BooleanType
```
Comment (ondrejmirtes) [inline src/Type/ComparisonHelper.php]: You've missed my point. This is supposed to be in `ConstantArrayType::looseCompare()`.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: src/Type/ComparisonHelper.php (hunk 1 above)
Comment (staabm) [reply src/Type/ComparisonHelper.php]: So should I duplicate the logic?

Its also called from resolveIdentical.

Comment (staabm) [reply src/Type/ComparisonHelper.php]: Or is the goal to also reduce resolveIdentical to just

isSuperTypeOf()->no() ?

Comment (ondrejmirtes) [reply src/Type/ComparisonHelper.php]: > Or is the goal to also reduce resolveIdentical to just

Yeah, probably, in the future. I'm fine if momentarily the code is a bit duplicated.

Comment (staabm) [reply src/Type/ComparisonHelper.php]: just duplicated the method and removed the callable arg to improve readability

**Comment by staabm** [comment]
should be good to go

**Comment by staabm** [comment]
rebased

**Pull request closed without merging**